src/Microsoft.AspNet.Security.OAuth/OAuthBearerAuthenticationHandler.cs

```diff
@@ -24,87 +35,167 @@ protected override AuthenticationTicket AuthenticateCore()
             return AuthenticateCoreAsync().GetAwaiter().GetResult();
         }
 
-        protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
+		/// <summary>
+		/// Searches the 'Authorization' header for a 'Bearer' token. If the 'Bearer' token is found, it is validated using <see cref="TokenValidationParameters"/> set in the options.
+		/// </summary>
+		/// <returns></returns>
+		protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
         {
+			ExceptionDispatchInfo authFailedEx = null;
+			OAuthRequestTokenContext requestTokenContext = null;
             try
             {
                 // Find token in default location
-                string requestToken = null;
-                string authorization = Request.Headers.Get("Authorization");
-                if (!string.IsNullOrEmpty(authorization))
+				requestTokenContext = new OAuthRequestTokenContext(Context, null);
+
+				// Give application opportunity to find from a different location, adjust, or reject token
+				var messageReceivedNotification =
+					new MessageReceivedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>(Context, Options)
+					{
+						ProtocolMessage = requestTokenContext,
+					};
+
+				// notification can set the token
+				await Options.Notifications.MessageReceived(messageReceivedNotification);
```

Comment:
Instead of raising `MessageReceived` event for every request in the pipeline it would be good if we raise this only if we know there is an `Authorization` header with value starting `Bearer`. Otherwise we will end up calling this too frequently. In the previous implementation we call this notification like that. In our katana's `OpenIDC` or `WSFed` implementation we raise the event only if the request matches several things like : 

1. Url's path is same as callback (there is no equivalent here). 
2. request is `POST` and containing a specific encoding etc.